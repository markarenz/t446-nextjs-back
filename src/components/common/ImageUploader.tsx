import { useState, useRef } from 'react';
import axios from 'axios';
import Button from '../common/Button';
import IconButton from './IconButton';
import { IconClose } from '../img/icons';
import { useAppContext } from '../../context/AppContext';
import styles from '../../styles/modules/ImageUploader.module.scss';

type StagedImage = {
  img: File;
  srcUrl: string;
};

const ImageUploader = () => {
  const [images, setImages] = useState<StagedImage[]>([]);
  const fileUploader = useRef<HTMLInputElement | null>(null);

  const uploadToClient = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target?.files && event.target?.files[0]) {
      const newImages: StagedImage[] = [];
      Object.values(event.target.files).forEach((i) => {
        if (!images.some((imgObj) => imgObj.img.name === i.name)) {
          newImages.push({
            img: i,
            srcUrl: URL.createObjectURL(i)
          });
        }
      });
      setImages([...images, ...newImages]);
    }
  };
  const handleRemoveImage = (imgName: string): void => {
    setImages(images.filter((imgObj) => imgObj.img.name !== imgName));
  };

  const uploadToServer = async (event: React.SyntheticEvent) => {
    event.preventDefault();
    const body = new FormData();
    // @ts-ignore
    body.append('files', images[0].img); //fileUploader?.current?.files[0]
    const response = await fetch('/api/assets/upload', {
      method: 'POST',
      body
    });
  };
  const isFormValid = images.length > 0;
  return (
    <div className={styles.uploader}>
      <form encType="multipart/form-data" action="">
        <div className={styles.uploaderStage}>
          <input
            type="file"
            ref={fileUploader}
            name="myImage"
            multiple
            accept="image/jpg, image/jpeg, image/png"
            onChange={uploadToClient}
            className={styles.imgUploadInput}
          />
          <div className={styles.messaging}>
            <h3>Drop images here or click to select.</h3>
            <Button
              onClick={uploadToServer}
              disabled={!isFormValid}
              color="primary"
            >
              <span>Upload</span>
            </Button>
          </div>
        </div>
      </form>
      <div className={styles.previews}>
        {images.map((imgObj) => (
          <div
            key={imgObj.img.name}
            className={styles.preview}
            style={{ backgroundImage: `url(${imgObj.srcUrl})` }}
          >
            <div className={styles.closeBtnWrap}>
              <IconButton
                title="Delete"
                color="primary"
                onClick={() => handleRemoveImage(imgObj.img.name)}
              >
                <IconClose />
              </IconButton>
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};

export default ImageUploader;
