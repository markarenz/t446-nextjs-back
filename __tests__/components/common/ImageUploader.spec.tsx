import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import ImageUploader from '../../../src/components/common/ImageUploader';
import { act } from 'react-dom/test-utils';

global.URL;
jest.mock('../../../src/context/AppContext', () => ({
  useAppContext: jest.fn().mockReturnValue({
    setLoading: jest.fn(),
    isSidebarOpen: true,
    setSidebarOpen: jest.fn()
  })
}));

global.fetch = jest
  .fn()
  .mockImplementationOnce(() =>
    Promise.resolve({ json: () => Promise.resolve({ filename: 'foobar.png' }) })
  )
  .mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve({ filename: 'foobar2.png' })
    })
  ) as jest.Mock;

let file: File;
let file2: File;
let file3: File;

beforeEach(() => {
  file = new File(['foo-bar'], 'foobar.png', { type: 'image/png' });
  file2 = new File(['foo2-bar2'], 'foobar2.png', { type: 'image/png' });
  file3 = new File(['foo-bar'], 'foobar.png', { type: 'image/png' });
  global.URL.createObjectURL = jest.fn();
});

const defaultProps = {
  handleTriggerRefresh: jest.fn()
};

describe('ImageUploader', () => {
  it('renders component', async () => {
    render(<ImageUploader {...defaultProps} />);
    const element = await screen.findByTestId('image-uploader');
    expect(element).toBeInTheDocument();
  });

  it('handles file input onchange & removal', async () => {
    render(<ImageUploader {...defaultProps} />);
    const uploader = await screen.findByTestId('file-upload-input');
    await waitFor(() =>
      fireEvent.change(uploader, {
        target: { files: [file] }
      })
    );
    await waitFor(
      () =>
        expect(screen.getByTestId('btn-upload-staged-0')).toBeInTheDocument(),
      {
        timeout: 2000
      }
    );
    const btnRemoveId = 'btn-upload-remove-0';
    await waitFor(
      () => expect(screen.getByTestId(btnRemoveId)).toBeInTheDocument(),
      {
        timeout: 2000
      }
    );
    const btnRemove = await screen.findByTestId(btnRemoveId);
    await waitFor(() =>
      fireEvent(
        btnRemove,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    const item = screen.queryByTestId('btn-upload-staged-0');
    expect(item).not.toBeInTheDocument();
  });

  it('calls fetch when uploading', async () => {
    render(<ImageUploader {...defaultProps} />);

    const uploader = await screen.findByTestId('file-upload-input');
    await waitFor(() =>
      fireEvent.change(uploader, {
        target: { files: [file, file2] }
      })
    );
    await waitFor(
      () =>
        expect(screen.getByTestId('btn-upload-staged-0')).toBeInTheDocument(),
      {
        timeout: 2000
      }
    );
    const btnUpload = await screen.findByTestId('btn-upload');
    await act(async () => {
      await waitFor(() =>
        fireEvent(
          btnUpload,
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true
          })
        )
      );
    });
    expect(global.fetch).toHaveBeenCalled();
  });

  it('allows only one file per filename', async () => {
    render(<ImageUploader {...defaultProps} />);

    const uploader = await screen.findByTestId('file-upload-input');
    await waitFor(() =>
      fireEvent.change(uploader, {
        target: { files: [file, file2] }
      })
    );
    expect(screen.queryByTestId('btn-upload-staged-0')).toBeInTheDocument();
    expect(screen.queryByTestId('btn-upload-staged-1')).toBeInTheDocument();
    await waitFor(() =>
      fireEvent.change(uploader, {
        target: { files: [file3] }
      })
    );
    expect(screen.queryByTestId('btn-upload-staged-2')).not.toBeInTheDocument();
  });
});
