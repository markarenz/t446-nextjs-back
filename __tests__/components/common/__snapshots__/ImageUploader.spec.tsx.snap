// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploader renders component 1`] = `
<div
  class="uploader"
  data-testid="image-uploader"
>
  <form
    action=""
    enctype="multipart/form-data"
  >
    <div
      class="uploaderStage"
    >
      <input
        accept="image/jpg, image/jpeg, image/png"
        class="imgUploadInput"
        data-testid="file-upload-input"
        multiple=""
        name="myImage"
        type="file"
      />
      <div
        class="messaging"
      >
        <h3>
          Drop images here or click to select.
        </h3>
        <button
          class="button  contained colorPrimary"
          data-testid="btn-upload"
          disabled=""
          role="button"
          type="button"
        >
          <span>
            Upload
          </span>
        </button>
      </div>
    </div>
  </form>
  <div
    class="previews"
  />
</div>
`;
